### ISSUE TITLE
sc_uint constructor from binary string initializes the wrong integer

### ISSUE THREAD

**Issue by MauroFoti** [accellera-official/systemc#74]
**SystemC version**: tested on 2.3.3, 2.3.4 and 3.0.0 (the latter pulled and compiled from main)
**Platform**: Linux Mint 20.3
**Compiler**: g++ 9.4.0

**Description of the problem**:

The `sc_uint` (and `sc_biguint`) constructor that takes a binary string as input seems to incorrectly initialize the number. The same number created from an integer number is initialized correctly.

The error seems to depend on the length of the integer, for example:

```cpp
cout << sc_uint<3>(2);      // prints 2
cout << sc_uint<3>("010");  // prints 2
cout << sc_uint<4>(2);      // prints 2
cout << sc_uint<4>("0010"); // prints 10
```

**Steps to reproduce the problem**

To verify the behavior, I've used the following code, that compares all `sc_uint` created from a binary string of length N and the corresponding integer, with an `sc_bv` created from the same values.

```cpp
#include "systemc.h"
#include <iostream>
#include <bitset> // used to print the binary string of an int

// how many bits to test
#define N 4

template <typename T>
void test(T n)
{
    sc_uint<N> integer(n); // create the integer
    sc_bv<N> bitvector(integer); // convert integer to bitvector
    
    // check if the converted number is the same as the original one
    cout << n << " " << bitvector << (n == bitvector ? " OK" : " ERROR") << endl;
}

int sc_main(int argc, char * argv[]) {

    // create sc_uint from integer
	for (int i = 0; i < (1 << N); i++)
	    test(i);

    // create sc_uint from binary string
	for (int i = 0; i < (1 << N); i++)
	    test(std::bitset<N>(i).to_string().c_str());

	return 0;
}
```

The results with N <= 3 report everything as correct

```
0 000 OK
1 001 OK
2 010 OK
3 011 OK
4 100 OK
5 101 OK
6 110 OK
7 111 OK
000 000 OK
001 001 OK
010 010 OK
011 011 OK
100 100 OK
101 101 OK
110 110 OK
111 111 OK
```

However, with N = 4, half of the numbers are wrong

```
0 0000 OK
1 0001 OK
2 0010 OK
3 0011 OK
4 0100 OK
5 0101 OK
6 0110 OK
7 0111 OK
8 1000 OK
9 1001 OK
10 1010 OK
11 1011 OK
12 1100 OK
13 1101 OK
14 1110 OK
15 1111 OK
0000 0000 OK
0001 0001 OK
0010 1010 ERROR
0011 1011 ERROR
0100 0100 OK
0101 0101 OK
0110 1110 ERROR
0111 1111 ERROR
1000 1000 OK
1001 1001 OK
1010 0010 ERROR
1011 0011 ERROR
1100 1100 OK
1101 1101 OK
1110 0110 ERROR
```

with N = 6, only 25% of numbers are correct, with N = 7 only 12.5% are correct and so on...

**Comment by maehne** [comment]
@AndrewGoodrich: Could you maybe have a look into this issue?

**Comment by pah** [comment]
Quoting the IEEE 1666-2023 SystemC standard, clause 7.3 (emphasis mine):

> A bit-pattern string (containing bit or logic character values **with no prefix**) **may be assigned to a vector**. If the number of characters in the bit-pattern string is less than the vector word length, the string shall be zero extended at its left-hand side to the vector word length. **The result of assigning such a string to a numeric type is undefined.**

Try adding a binary prefix to your strings, e.g. `0b0010`.

(The current behaviour of silently producing surprising results is another question.)

**Comment by MauroFoti** [comment]
> Try adding a binary prefix to your strings, e.g. `0b0010`.
> 
> (The current behaviour of silently producing surprising results is another question.)

Adding the prefix "0b" correctly creates the integers correctly

**Comment by AndrewGoodrich** [comment]
When assigning string values to SystemC integer data types a local fixed point variable is created as an intermediary (this goes all the way back to the original Synopsys implementation). 

In the case of:
    cout << sc_uint<3>("010");
and
    cout << sc_uint<4>("0010");
The code will filter through scfx_rep::from_string(const char*), which will call scfx_parse_prefix(). scfx_parse_prefix() looks for an initial zero followed by a base designation. If that is not found it sets the base type to decimal. This explains why "0010" returns a value of 10.  The code will eventually flow through sc_uint_base::operator = ( const sc_fxnum& v ), where the value of v is decimal 10.

In the case of "010" we have a similar flow, but when we hit sc_uint_base::operator = ( const sc_fxnum& v ) the value of v is decimal 2! The difference in the flow is a call to scfx_rep::o_extend() in the "010" case. The code in o_extend() is:

inline
void
scfx_rep::o_extend( const scfx_index& x, sc_enc enc )
{   
    int wi = x.wi();
    int bi = x.bi();
    
    SC_ASSERT_( wi >= 0 && wi < size(), "word index out of range" );
    
    if( enc == SC_US_ || ( m_mant[wi] & ( ((word)1) << bi ) ) == 0 )
    {
        if( bi != bits_in_word - 1 )
            m_mant[wi] &= ~( ((word)-1) << ( bi + 1 ) );
        for( int i = wi + 1; i < size(); ++ i )
            m_mant[i] = 0;
        m_sign = 1;
    }
    else 
    {
        if( bi != bits_in_word - 1 )
            m_mant[wi] |= ( ((word)-1) << ( bi + 1 ) );
        for( int i = wi + 1; i < size(); ++ i )
            m_mant[i] = static_cast<word>( -1 );
        m_sign = -1;
    }
}   

The value of bi is 2 so m_mant[0],  whose value is 10 (0xa), will be anded with 0x7 resulting in a value of 2. So "010" has  the problem, not "0010". With:

   cout << sc_uint<3>("011");

a value of 3 is set, and it would have been apparent that sc_uint<3>("010")  was not treating "010" as a 2, but rather truncating its value.

I need to track down a couple of things:

    (1) Why o_extend() gets called in one case and not the other.
    (2) Why bi is two rather than 4

 I have always hated the FX code!

**Comment by MauroFoti** [comment]
@pah  According to your suggestion, and to @AndrewGoodrich description of the flow, without "0b", the string should always be interpreted as decimal. I did some debugging, and the base is always correctly parsed inside `scfx_parse_prefix` (`SC_DEC` with "0100" and `SC_BIN` with "0b0100"). Also, inside `scfx_rep::from_string(const char*)` the mantissa is correctly parsed, which means the intermediate fixed number is created correctly.

@AndrewGoodrich  `o_extend()` is called in the overflow function inside `scfx_rep.cpp`, which actually makes sense considering that I was trying to create a 4-bit integer with a 4-digit decimal number as initializer, which will almost never fit on lower values of N. I tried changing my test program by increasing the number of bits of the values and the results are always correct if I parse the string as base 10

```
0 0 000000000000000 OK
1 1 000000000000001 OK
2 2 000000000000010 OK
3 3 000000000000011 OK
4 4 000000000000100 OK
5 5 000000000000101 OK
6 6 000000000000110 OK
7 7 000000000000111 OK
000000000000000 0 000000000000000 OK
000000000000001 1 000000000000001 OK
000000000000010 10 000000000001010 ERROR
000000000000011 11 000000000001011 ERROR
000000000000100 100 000000001100100 ERROR
000000000000101 101 000000001100101 ERROR
000000000000110 110 000000001101110 ERROR
000000000000111 111 000000001101111 ERROR
```

The "surprising results" are actually just a coincidence where the overflowing number, once truncated, actually matches the binary representation of the original value: decimal 010 (0b1010), when and'd with 0b111, actually returns 0b010 = 2, and, quite curiously, this is actually true for all the 3 (decimal digits) numbers and their 3-bit representation. When going above N=3, the overflow, truncated number is different from its binary representation and inconsistencies start appearing.

For example, with N=4, decimal 1000 & 0b1111 = 0b1000 (decimal 8), while decimal 1010 & 0b1111 = 0b0010 (decimal 2).

From all of this, I would say that there is no bug, the code is working as expected, I just ignored that to initialize a binary string a base prefix was needed, and this curious mathematical coincidence hid my mistake in some cases. However, the constructor also silently performed a truncation because of overflow without raising any warning, is that intended behavior?
